### PR TITLE
Add an operator [ paictl conf update ] to upload configuration to k8s 

### PR DESCRIPTION
**Pull request by ydye**
#### Progress
- [x] paictl config update -p /path/to/local/configuration [-c kubeconfig-path ]
- [x] paictl config update -e external-storage-config [-c kubeconfig-path]
- [x] paictl config update [-c kubeconfig-path]

image

- [x] paiclt.py config external-config-update -e external-configuration [-c kube-config]

image

- [x] paictl.py config get -o /path/to/output [-c kube-config]

#### Change
1) With the provided external-configuration-storage conf, you could get your cluster-configuration from the service like git, and then upload it to k8s cluster.
2) The cluster-configuration will be updated into a configmap. And all service in the k8s cluster could get it from configmap volume.
3) Tool ```paictl``` will get the configuration from k8s-configmap. So next time, admin should update their cluster configuration from external storage to the cluster. Then manage the openpai cluster.

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -64,6 +64,9 @@ sudo chown core:core -R /pathHadoop/
 QUICK_START_PATH=${JENKINS_HOME}/${BED}/singlebox/quick-start
 CONFIG_PATH=${JENKINS_HOME}/${BED}/singlebox/cluster-configuration
 
+# Install necessary python package
+sudo pip install GitPython
```
Comment (hao1939) [inline Jenkinsfile]: Could we remove the line? Noticed that you have updated the dependencies in `dev-box`.

Comment (ydye) [reply Jenkinsfile]: No we can't. paictl need import it. You will call paictl to generate cluster-configuration out of the dev-box.

**Comment by ydye** [comment]
#1440

File: deployment/confStorage/external_version_control/git_storage.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+            self.logger.error("Path: {0}".format(self.path))
+            self.logger.error("Repo: {0}".format(self.repo_url))
+            sys.exit(1)
+
+        pai_temp_path = "./tmp-config-{0}".format(self.time)
+        directory_handler.directory_copy("{0}/*".format(configuation_path), pai_temp_path)
+
+
+
+    def open(self):
```
Comment (xudifsd) [inline deployment/confStorage/external_version_control/git_storage.py]: maybe you can use `__enter__` and `__exit__`, which is more elegant, and can clean storage in exit.

Same applies to `local_storage`.

File: deployment/confStorage/external_version_control/git_storage.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+    def __init__(self, storage_configuration, local_store = "/tmp/pai-conf-sync/git-storage"):
+
+        self.repo_url = storage_configuration["url"]
+        self.branch = storage_configuration["branch"]
+        self.path = storage_configuration["path"]
+
+        self.time = str(int(time.time()))
+        self.logger = logging.getLogger(__name__)
+        self.local_store = "{0}-{1}".format(local_store, self.time)
```
Comment (xudifsd) [inline deployment/confStorage/external_version_control/git_storage.py]: why not using [`tempfile`](https://docs.python.org/2/library/tempfile.html)

File: deployment/confStorage/external_version_control/git_storage.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+
+
+    def git_clone(self):
+        try:
+            self.repo = git.Repo.clone_from(self.repo_url, self.local_store, branch=self.branch)
```
Comment (xudifsd) [inline deployment/confStorage/external_version_control/git_storage.py]: can we just clone with `--depth=1`, since we requires no history.

Comment (ydye) [reply deployment/confStorage/external_version_control/git_storage.py]: Changed.

File: deployment/confStorage/external_version_control/git_storage.py (hunk 2 above)
Comment (ydye) [reply deployment/confStorage/external_version_control/git_storage.py]: Changed

File: deployment/confStorage/external_version_control/git_storage.py (hunk 3 above)
Comment (ydye) [reply deployment/confStorage/external_version_control/git_storage.py]: We will have to remove the temp directory created by ```tempfile.mkdtemp([suffix=''[, prefix='tmp'[, dir=None]]]) ```.  And I think it's same as what we have done.

**Pull request merged**